### PR TITLE
410 Integral Set 

### PR DESCRIPTION
**Pull request by lifflander**
- Implemented new integral set vt::IntegralSet, that makes epoch alloc/dealloc far more efficient
    - Uses a sorted, non-overlapping interval tree to implement automatic compaction
    - Holes in the release set of epochs no longer inhibit wrap around (a major problem for tree-structured epochs)

### CODE REVIEW

File: src/vt/termination/interval/interval.h [hunk 1]
```diff
@@ -0,0 +1,226 @@
...
+   *
+   *        TangentLeft                   TangentRight
+   *       |------i-----| lb_        ub_ |------i-----|
+   *                      |----this----|
+   *                     ^              ^
+   *                    gap            gap
+   */
+
+  PositionEnum tangent(IntervalType const& i, DomainT gap = 1) const {
+    if (lb_ == i.ub_ + gap) {
```
Comment (PhilMiller) [inline src/vt/termination/interval/interval.h]: The logic of the tangent being exactly `gap` away doesn't match the comment of "within a specified gap". Without a use case, I can't tell which is right.

File: src/vt/termination/term_window.cc [hunk 2]
```diff
@@ -79,148 +75,68 @@ inline bool EpochWindow::isArchetypal(EpochType const& epoch) {
...
-    /*
-     * Set the first and last epoch, since this is the first time this window
-     * is being initialized for the given epoch configuration
-     */
-    first_unresolved_epoch_ = last_unresolved_epoch_ = epoch;
-  } else {
-    last_unresolved_epoch_ = std::max(last_unresolved_epoch_, epoch);
+  // We should possibly perform some error checking once we wrap around in case
+  // of pending actions
+  if (terminated_epochs_.contains(epoch)) {
```
Comment (PhilMiller) [inline src/vt/termination/term_window.cc]: I think a possible assertion would be

`epoch > getLast() || terminated_epochs_.contains(epoch)`

I.e. either we haven't wrapped around, or we have, and we're clear to re-use this value. We run into trouble with that when `terminated_epochs_` itself covers a wrapped-around range, though. I.e. [0,100] and [MAX-100, MAX].

**Comment by PhilMiller** [comment]
Looks good enough overall. I'm still uneasy about the unused `gap` parameter and `set` methods, but they're only a small distraction, not breaking anything present.

**Comment by lifflander** [comment]
@PhilMiller Ok, will you approve this? I want to get this merged. I'm tried of rebasing it!

I changed the settings at least one other person has to approve before github will allow merging

**Pull request merged**